larq/quantizers.py

```diff
@@ -601,6 +603,86 @@ def get_config(self):
+    w_{lim}(w) = \tanh(w)
+    \\]
+
+    Furthermore, the weights of each layer are normed, such that the weight with
+    the largest magnitude gets the largest or smallest (depending on its sign)
+    quantizable value. That way, the full quantizable numeric range is utilized.
+
+    \\[
+    w_{norm}(w) = \frac{w}{\max(|w|)}
+    \\]
+
+    The formulas can be found in the paper in section 2.3. Please note, that
+    the paper refers to weights being quantized on a numeric range of [-1, 1], while
+    activations are quantized on the numeric range [0, 1]. `DoReFa` defines the
+    quantization function quantizek() from the paper with the correct numeric
+    range of [0, 1], which is why the range needs to be adapted before applying
+    `DoReFa` after applying the preprocessing (limit and norm).
+    The hard limiting inside `DoReFa` becomes ineffective, because its input
+    is already limited by the hyperbolic tangent. The full quantization
+    function including the adaption of numeric ranges is
+
+    \\[
+    q(w) = 2 \, quantize_{k}(\frac{w_{norm}\left(w_{lim}\left(w\right)\right)}{2} + \frac{1}{2}) - 1
+    \\]
+
+    !!! warning
+        This quantizer works for weights on the range [-1, 1], which matches the
+        default setting of `constraints.weight_clip`. Do not use this quantizer
+        with a different constraint `clip_value` than the default one.
+
+    ```plot-activation
+    quantizers.DoReFaKernel
+    ```
+
+    The interface of this quantizer is the same like the one of `DoReFa`.
+
+    # References
+        - [DoReFa-Net: Training Low Bitwidth Convolutional Neural Networks with Low
+            Bitwidth Gradients](https://arxiv.org/abs/1606.06160)
+    """
+
+    def call(self, inputs):
+
+        limited = tf.math.tanh(inputs)
+
+        # Divider for max-value norm.
+        dividend = tf.math.reduce_max(tf.math.abs(limited))
+
+        # Need to stop the gradient here. Otherwise, for the maximum element,
+        # which gives the dividend, normed is limited/limited (for this one
+        # maximum digit). The derivative of y = x/x, dy/dx is just zero, when
+        # one does the simplification y = x/x = 1. But TF does NOT do this
+        # simplification when computing the gradient for the
+        # normed = limited/dividend operation. As a result, this gradient
+        # becomes complicated, because during the computation, "dividend" is
+        # not just a constant, but depends on "limited" instead. Here,
+        # tf.stop_gradient is used to mark "dividend" as a constant explicitly.
+        dividend = tf.stop_gradient(dividend)
+
+        # Norm and then scale from value range [-1,1] to [0,1].
+        # If the dividend used for the norm operation is 0, all elements of
+        # the weight tensor are 0 and divide_no_nan returns 0 for all weights.
+        # So if all elements of the weight tensor are zero, nothing is normed.
+        normed = tf.math.divide_no_nan(limited, dividend) / 2.0 + 0.5
```

Comment:
```suggestion
        normed = tf.math.divide_no_nan(limited, 2.0 * dividend) + 0.5
```
This keeps the notation closer to the formula in the paper, if you don't mind